Spring Data: Fix interface projection using Page and Slice

Sgitario opened the pull request:
The interface project worked fine using Lists, but not using Page and Slice. Example:

```java
@Repository
public interface BasicTypeDataRepository extends CrudRepository<BasicTypeData, Integer> {

    // works
    @Query(value = "select id, doubleValue from BasicTypeData")
    List<WithDoubleValue> findAllDoubleValuesToList();

   // it does not work (1)
    @Query(value = "select id, doubleValue from BasicTypeData")
    Page<WithDoubleValue> findAllDoubleValuesToPage();
}
```

This PR fixes the interface projection when returning a Page object (1).

Fixes https://github.com/quarkusio/quarkus/issues/21616

famod commented:
I never worked with it, but what about `Slice`? `Page` extends `Slice`, btw...

Sgitario commented:
> I never worked with it, but what about `Slice`? `Page` extends `Slice`, btw...

Well spotted! I've already fixed it for Slice too. Thanks.

famod commented on extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java:
```diff
@@ -75,6 +79,33 @@ public void testTimeZonesByLocale() {
+    @Order(6)
+    @Transactional
+    public void testMapInterfacesUsingPage() {
+        Page<WithDoubleValue> list = repo.findAllDoubleValuesToPage();
```
Nitpick: `list` variable name is a bit misleading. Same for the `Slice` test method.

Sgitario replied:
Changed

famod commented on extensions/spring-data-jpa/deployment/src/test/java/io/quarkus/spring/data/deployment/BasicTypeDataRepositoryTest.java:
```diff
@@ -75,6 +79,33 @@ public void testTimeZonesByLocale() {
+        Page<WithDoubleValue> list = repo.findAllDoubleValuesToPage();
+        Set<Double> actual = list.stream().map(WithDoubleValue::getDoubleValue).collect(Collectors.toSet());
+        assertThat(actual).isNotEmpty().contains(Math.PI);
+    }
```
What about testing `getTotalElements()` etc.? Same for the `Slice` test method.

Sgitario replied:
Done

famod reviewed (approved):
LGTM, only two small remarks/suggestions.

geoand reviewed (commented):
Thanks!

@famod covered me on this :)

Sgitario committed:
Spring Data: Fix interface projection using Page and Slice

The interface project worked fine using Lists, but not using Page or Slice. Example:

```java
@Repository
public interface BasicTypeDataRepository extends CrudRepository<BasicTypeData, Integer> {

    // works
    @Query(value = "select id, doubleValue from BasicTypeData")
    List<WithDoubleValue> findAllDoubleValuesToList();

   // it does not work (1)
    @Query(value = "select id, doubleValue from BasicTypeData")
    Page<WithDoubleValue> findAllDoubleValuesToPage();
}
```

This PR fixes the interface projection when returning a Page or a Slice object (1).

Fixes https://github.com/quarkusio/quarkus/issues/21616

Sgitario commented:
Updated PR to cover some functionality around `Pageable`.
